lib/6to5/traverse/index.js

```diff
@@ -50,104 +57,139 @@ TraversalContext.prototype.maybeReplace = function (result, obj, key, node) {
   }
 
   if (isArray) {
-    this.flatten();
+    return true;
+  }
+}
+
+TraversalContext.prototype.enterNode = function (obj, key, node, enter, parent, scope, state) {
+  var result = enter(node, parent, scope, this, state);
+  var flatten = false;
+
+  if (result) {
+    flatten = replaceNode(obj, key, node, result);
+    node = result;
+
+    if (flatten) {
+      this.didFlatten = true;
+    }
+  }
+
+  if (this.didRemove) {
+    obj[key] = null;
+    this.didFlatten = true;
+  }
+
+  return node;
+};
+
+TraversalContext.prototype.exitNode = function (obj, key, node, exit, parent, scope, state) {
+  var result = exit(node, parent, scope, this, state);
+  var flatten = false;
+
+  if (result) {
+    flatten = replaceNode(obj, key, node, result);
+    node = result;
+
+    if (flatten) {
+      this.didFlatten = true;
+    }
   }
 
   return node;
 };
 
-TraversalContext.prototype.visit = function (obj, key, opts, scope, parent) {
+TraversalContext.prototype.visitNode = function (obj, key, opts, scope, parent, state) {
+  this.reset();
+
   var node = obj[key];
-  if (!node) return;
 
   // type is blacklisted
-  if (opts.blacklist && opts.blacklist.indexOf(node.type) > -1) return;
+  if (opts.blacklist && opts.blacklist.indexOf(node.type) > -1)
```

Comment:
so, I was curious.

`node  v0.11.14`

https://static.iamstef.net/index-of

property hit (interned string): 
* `fastObject` is always faster then `indexOf`
* `object` overtakes `indexOf` when a scan of 11 entries is required

property miss (interned string): (this one is obviously much more volatile)
* `indexOf` is faster up until a scan of > 11 entries is required, the `null object` surpasses
* `null object` > `null fast object` > `fastObject`

https://static.iamstef.net/index-of